_posts/2020-05-11-microservices-on-axon.md

```diff
@@ -0,0 +1,198 @@
+
+Potencjalnym rozwiązaniem pierwszego problemu mógłby być osobny topic dla każdego agregatu, wówczas odpada konieczność filtrowania eventów.
+To rozwiązanie jednak może nie sprawdzić się przy ogromnej ilości agregatów. 
+Wynika to ze sposobu, w jaki Kafka przechowuje topici (a właściwie partycje) - dla każdej tworzony jest osobny katalog w systemie plików. 
+Szczegółowe wyjaśnienie znajduje się w [**filmie**](https://youtu.be/zUSWsJteRfw?t=2179) przygotowanym przez AxonIQ (firma odpowiedzialna za rozwój Axona).
+
+### AxonServer
+W kwestii event store AxonIQ wyszedł na przeciw potrzebom dając do dyspozycji swoje narzędzie, które idealnie spełnia się w tej roli - AxonServer:
+- Pozwala na dokładanie eventów (z jednoczesnym brakiem możliwości edycji już istniejących).
+- Zapewnia stałą wydajność niezależnie od ilości danych przetrzymywanych w event store.
+- Umożliwia konstruowanie snapshotów dla agregatów i nakładanie ich (w przypadku dużej ilości eventów rekonstrukcja agregatu bez funkcjonalności snapshotów może trochę trwać).
+
+Po uruchomieniu AxonServera mamy dostęp do dashboardu pokazującego, który mikroserwis jest podpięty pod event store wraz z jego liczbą instancji:
+![AxonDashboard](/assets/img/posts/2020-05-11-microservices-on-axon/axon_dashboard.png)
+Na samym dashboardzie, funkcjonalności panelu administracyjnego się nie kończą:
+- Podgląd konfiguracji wraz z przepustowością (commandy/eventy/query/snapshoty na sekundę).
+- Możliwość wyszukiwania eventu przy użyciu zapytań.
+- Tabelka ze wskazaniem, który command, ile razy i w jakim serwisie został obsłużony.
+- Zarządzanie dostępem do panelu.
+
+Oczywiście AxonFramework jest w pełni kompatybilny z AxonServerem i działa out-of-the-box, bez dodatkowej konfiguracji.
+
+# Najpierw monolit
+Zaczynając przygodę z Axonem, nie chciałem skakać na głęboką wodę, zacząłem więc od monolitu, mając jednak z tyłu głowy perspektywę zmigrowania na coś bardziej skalowalnego.
+Migracja z monolitu na mikroserwisy nierzadko sprawia wiele problemów, tak było również w moim przypadku z [**tą aplikacją**](https://github.com/matty-matt/movie-keeper-core).
+W skrócie pozwala ona na wyszukiwanie filmów po tytułach, wraz z ich obsadą oraz trailerami korzystając z [**API TMDb**](https://developers.themoviedb.org/3/getting-started), zapisywanie wszystkiego w bazie, oznaczanie filmu jako przeczytany oraz sprawdzanie premiery cyfrowego wydania.
+Stworzyłem więc agregat filmu wraz z encjami zawierającymi trailery oraz obsadę:
+```java
+@Aggregate
+public class MovieAggregate {
+    @AggregateIdentifier
+    private MovieId movieId;
+    @AggregateMember
+    private TrailerEntity trailerEntity;
+    @AggregateMember
+    private CastEntity castEntity;
+    ...
+}
+```
+Pobieranie danych z zewnętrznego serwisu działo się w EventHandlerze, poprzez zawołanie odpowiedniej metody z interfejsu ExternalService:
+```java
+@Service
+public class MovieEventsHandler {
```

Comment:
co to za zdarzenia? początkowe zasilenie danymi? coś co częściej rzucasz? MoveSearchDelegatedEvent to zdarzenie domenowe czy po prosty aplikacyjne? :)

Reply:
MovieEventsHandler robił mi za klaskę pośredniczącą do external serwisu, czyli każde wyszukanie nowego filmu wchodziło tutaj. Jeżeli chodzi o MoveSearchDelegatedEvent to jest to w sumie zdarzenie aplikacyjne - wg koncepcji DDD commandy i eventy zdaje się że powinny być domenowe, ale nie zawsze się tak da :P

Reply:
Ale to nie robi gdzieś zdarzeń pod spodem? (nie widzę przepływu z SaveMovieCommand)? Bo chyba nie jest tak, że co query odkłada się zdarzenie? ;)

Reply:
Hmm nie do końca rozumiem pytania. Przepływ wygląda mniej więcej tak:
- HTTP POST z tytułem filmu, który wysyła FindMovieCommand oraz zakłada subskrypcję na GetMovieQuery (potrzebuję to aby od razu zwrócić wynik)
- Agregat łyka command i emituje MovieSearchDelegatedEvent
- MovieSearchDelegatedEvent handlowany jest w powyższej klasce i w zależności od rezultatu albo emituje NOT_FOUND do subscriptionQuery (i zwracam 404), albo wysyła SaveMovieCommand (który emituje kolejne eventy i zapisuje film do mongo, wyszukuje trailery cast itd)

W momencie gdy chcemy pobrać film po id dzieje się to też z GetMovieQuery, al ejuż bez subskrypcji.

Reply:
W razie co cały kod monolitu sprzed migracji mam tu:
https://github.com/matty-matt/movie-keeper-core/tree/monolith